interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,405 @@
+// prompted.
+type Prompt struct {
+	ID           string             `json:"id"`
+	Timestamp    time.Time          `json:"timestamp"`
+	Snap         string             `json:"snap"`
+	Interface    string             `json:"interface"`
+	Constraints  *PromptConstraints `json:"constraints"`
+	listenerReqs []*listener.Request
+}
+
+// PromptConstraints are like prompting.Constraints, but have a "path" field
+// instead of a "path-pattern", and include the available permissions for the
+// interface corresponding to the prompt.
+type PromptConstraints struct {
+	Path                 string   `json:"path"`
+	Permissions          []string `json:"permissions"`
+	AvailablePermissions []string `json:"available-permissions"`
+}
+
+// equals returns true if the two prompt constraints are identical.
+func (pc *PromptConstraints) equals(other *PromptConstraints) bool {
+	if pc.Path != other.Path || len(pc.Permissions) != len(other.Permissions) {
+		return false
+	}
+	// Avoid using reflect.DeepEquals to compare []string contents
+	for i := range pc.Permissions {
+		if pc.Permissions[i] != other.Permissions[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// subtractPermissions removes all of the given permissions from the list of
+// permissions in the constraints.
+func (pc *PromptConstraints) subtractPermissions(permissions []string) (modified bool) {
+	newPermissions := make([]string, 0, len(pc.Permissions))
+	for _, perm := range pc.Permissions {
+		if !strutil.ListContains(permissions, perm) {
+			newPermissions = append(newPermissions, perm)
+		}
+	}
+	if len(newPermissions) != len(pc.Permissions) {
+		pc.Permissions = newPermissions
+		return true
+	}
+	return false
+}
+
+// userPromptDB maps prompt IDs to prompts for a single user.
+type userPromptDB struct {
+	ByID map[string]*Prompt
+}
+
+// PromptDB stores outstanding prompts in memory and ensures that new prompts
+// are created with a unique ID.
+type PromptDB struct {
+	// Mmap the max ID file to avoid unnecessary syscalls
+	maxIDMmap []byte
+	// The per-user DB is protected by a RWMutex
+	mutex   sync.RWMutex
+	perUser map[uint32]*userPromptDB
+	// Function to issue a notice for a change in a prompt
+	notifyPrompt func(userID uint32, promptID string, data map[string]string) error
+}
+
+const (
+	maxIDFileSize                int = 8
+	maxOutstandingPromptsPerUser int = 1000
+)
+
+// New creates and returns a new prompt database.
+//
+// The given notifyPrompt closure should record a notice of type
+// "interfaces-requests-prompt" for the given user with the given
+// promptID as its key.
+func New(notifyPrompt func(userID uint32, promptID string, data map[string]string) error) (*PromptDB, error) {
+	pdb := PromptDB{
+		perUser:      make(map[uint32]*userPromptDB),
+		notifyPrompt: notifyPrompt,
+	}
+	maxIDFilepath := filepath.Join(dirs.SnapRunDir, "request-prompt-max-id")
+	maxIDFile, err := os.OpenFile(maxIDFilepath, os.O_RDWR, 0600)
```

Comment:
since the `else` branch checks the file size anyway, can't we use `os.O_RDWR|os.O_CREATE` and simply initialize it when it's been created?

Reply:
Great point, that makes things much simpler.